### PR TITLE
Fix #4172: Expose ride-type changing capability in UI

### LINKED ISSUE #4172: Expose ride-type changing capability in UI
As a follow-up to #3273, we could expose ride-type-changing functionality in the UI as another options under cheats. We already do similar thing by allowing different vehicle types.

It may be as easy as adding it to `ride_set_setting` in here: https://github.com/OpenRCT2/OpenRCT2/blob/2d1cbc916fe7f3400e55920cf57a4ad7a28e3d36/src/ride/ride.c#L3783 and wiring things up through `game_command_set_ride_setting`.

Doing it this way would also let it sync across multiplayer.

### PR DESCRIPTION
**Pull request by janisozaur**
This also allows changing ride type in multiplayer games. Very crashy.

### CODE REVIEW

**Comment by janisozaur** [comment]
Please review, but I'd like to take this opportunity to up the network version number to make one more change, so don't merge just yet (highly unlikely it's merge quality already :P)

**Comment by janisozaur** [comment]
OK, I'm done with all the things I wanted.

**Comment by janisozaur** [comment]
As this PR modifies network version stream, it would be best to merge https://github.com/OpenRCT2/OpenRCT2/pull/4173 first to make sure everyone on new version has it.

**Comment by IntelOrca** [comment]
There some invalidation issues:
* When switching to the first tab on the ride, the viewport doesn't redraw instantly.
* When changing the ride type, it doesn't change instantly because its not invalidating the widget.
* When applying the ride type, it should invalidate the whole screen.

There is also not enough space between the ride type spinner and the bottom of the window. You should try to maintain an equal margin around the edge of the window.

**Comment by janisozaur** [comment]
@IntelOrca latest commit should address all the issues you mentioned. I have also fixed mingw builds, it was a case of bad cache on travis.

File: src/windows/ride.c [hunk 1]
```diff
@@ -2247,6 +2291,7 @@ static void window_ride_main_invalidate(rct_window *w)
 
 	set_format_arg(0, uint16, ride->name);
 	set_format_arg(2, uint32, ride->name_arguments);
+	set_format_arg(6, uint16, _rideType + 2);
```
Comment (marijnvdwerf) [inline src/windows/ride.c]:
```c
set_format_arg(6, rct_string_id, _rideType + STR_RIDE_NAME_SPIRAL_ROLLER_COASTER);
```

or one #4097 is merged:
```c
set_format_arg(6, rct_string_id, RideNaming[_rideType].name _ride);
```

**Comment by marijnvdwerf** [comment]
An 'Apply' button seems rather inconsistent with the current UI. Would make more sense to put it in a dialog, if it's required.

**Comment by janisozaur** [comment]
@marijnvdwerf I am **not** doing any dialogs with our current UI-handling code.

**Comment by janisozaur** [comment]
@marijnvdwerf I have updated string name

@IntelOrca this was rebased on top of #4175 now.

**Comment by marijnvdwerf** [comment]
What about a dropdown instead? That could require two clicks as well + is a less sucky experience than a spinner.

File: src/localisation/string_ids.h [hunk 2]
```diff
@@ -3336,6 +3336,10 @@ enum {
 	STR_DEBUG_PAINT_SHOW_BOUND_BOXES = 5902,
 	STR_DEBUG_DROPDOWN_DEBUG_PAINT = 5903,
 
+	STR_CHEAT_ALLOW_ARBITRARY_RIDE_TYPE_CHANGES = 5904,
+	STR_CHEAT_ALLOW_ARBITRARY_RIDE_TYPE_CHANGES_TIP = 5905,
+	STR_APPLY = 5906,
```
Comment (marijnvdwerf) [inline src/localisation/string_ids.h]: `STR_RIDE_TYPE_APPLY` or something along the lines?

Comment (janisozaur) [inline src/localisation/string_ids.h]: It's a generic `apply`

**Comment by janisozaur** [comment]
@marijnvdwerf I would prefer drop down as well, but they only handle 64 entries. You're welcome to fix that.

**Comment by marijnvdwerf** [comment]
Won't the game crash when you accidentally change the ride type to one that doesn't have as high of a track element type as you're using? If so: seems odd to have such an option visible in the UI; even when enabled by a cheat. Actually it sounds like something that would be more fit for the tile inspector.

**Comment by janisozaur** [comment]
Yes, it may crash (actually: it may case the game to crash or the ride to crash). We do already have "show vehicles from other tracks types" which used to be just as much crashy as this is.

As a middle-ground, I can split this commit in two: one would be logic changes and `rides set` handling, the other would be UI changes, which could possibly be merged at a later time.

@IntelOrca @Gymnasiast what're your thoughts on that?

**Comment by IntelOrca** [comment]
You seem to have a lot of code for multiplayer and sprite checking, shouldn't that be a separate PR?

> Won't the game crash when you accidentally change the ride type to one that doesn't have as high of a track element type as you're using?

We could use this to fix the crashes. Even though unintended we can make our code more defensive against hacks.

**Comment by janisozaur** [comment]
> You seem to have a lot of code for multiplayer and sprite checking, shouldn't that be a separate PR?

Ideally yes, but this changes network stream and requires a version bump, just like the cheat. I've decided to coalesce these two items to avoid having to bump network version twice, just like we did with https://github.com/OpenRCT2/OpenRCT2/pull/3782

I mostly mean how the checksum got added to `NETWORK_COMMAND_TICK` packet. I can split it in two, if you want it that way. I may even be able to do without requiring a network version update now that I think of it, if `NetworkPacket` offers a way to sense its size or returns `0` when trying to `>>`-read non-existent value.

**Comment by IntelOrca** [comment]
> if NetworkPacket offers a way to sense its size or returns 0 when trying to >>-read non-existent value

All of it is protected against reading past end of data buffer. It returns 0 for all bytes past it.

**Comment by janisozaur** [comment]
@IntelOrca I have extracted sprite-checksumming commit into its own PR: https://github.com/OpenRCT2/OpenRCT2/pull/4179

**Comment by duncanspumpkin** [comment]
It will be crashing in original code though so we can't prevent the crashes. I think cs changed how many track elem types there were for rides half way through development. This meant he only defined draw arrays of X amount for some and Y amount for others. When a ride with X amount tries to draw a track elem from the Y area it will crash but if it is inside of the x area and it doesn't have a function it correctly no ops. We can't just define the complete arrays with the correct null elements as the code would need a modified callproc due to the way paint setup code works.

**Comment by IntelOrca** [comment]
@janisozaur can you rebase this? I will take another look and then merge.

**Comment by janisozaur** [comment]
I have added @zsilencer's multiplayer fixes into this PR, so they get enabled for new network version.

File: src/windows/ride.c [hunk 3]
```diff
@@ -2360,15 +2405,46 @@ static void window_ride_main_invalidate(rct_window *w)
...
+	window_ride_main_widgets[WIDX_RIDE_TYPE_DECREASE].left = w->width - 98;
+	window_ride_main_widgets[WIDX_RIDE_TYPE_DECREASE].top = w->height - 10;
+	window_ride_main_widgets[WIDX_RIDE_TYPE_DECREASE].bottom = w->height - 5;
+	window_ride_main_widgets[WIDX_RIDE_TYPE_APPLY].left = w->width - 83;
+	window_ride_main_widgets[WIDX_RIDE_TYPE_APPLY].right = w->width - 26;
+	window_ride_main_widgets[WIDX_RIDE_TYPE_APPLY].top = w->height - 17;
+	window_ride_main_widgets[WIDX_RIDE_TYPE_APPLY].bottom = w->height - 4;
+
+	widget_invalidate(w, WIDX_VIEWPORT);
+	widget_invalidate(w, WIDX_VIEW);
```
Comment (IntelOrca) [inline src/windows/ride.c]: This is a large performance hit, you should only invalidate things when necessary.

**Pull request merged**